Support device map for distributed autograd while using TensorPipe.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44859 Support device map for distributed autograd while using TensorPipe.**

TensorPipe's `set_device_map` option was applied during the forward
pass. However, if we ran the backward pass for the graph we would not
automatically pick up the reverse device mapping.

As a result, users had to specify both forward and backward device mapping
which is very tedious to do.

In this PR, I've added this functionality such that TensorPipe automatically
picks up the reverse device mapping during the backward pass. This is done by
storing the appropriate device mapping in the "recv" autograd function for
distributed autograd.

#Closes: https://github.com/pytorch/pytorch/issues/44170

Differential Revision: [D23751975](https://our.internmc.facebook.com/intern/diff/D23751975/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23751975/)!

pritamdamania committed:
Support device map for distributed autograd while using TensorPipe.

TensorPipe's `set_device_map` option was applied during the forward
pass. However, if we ran the backward pass for the graph we would not
automatically pick up the reverse device mapping.

As a result, users had to specify both forward and backward device mapping
which is very tedious to do.

In this PR, I've added this functionality such that TensorPipe automatically
picks up the reverse device mapping during the backward pass. This is done by
storing the appropriate device mapping in the "recv" autograd function for
distributed autograd.

#Closes: https://github.com/pytorch/pytorch/issues/44170

Differential Revision: [D23751975](https://our.internmc.facebook.com/intern/diff/D23751975/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23751975/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Support device map for distributed autograd while using TensorPipe."


TensorPipe's `set_device_map` option was applied during the forward
pass. However, if we ran the backward pass for the graph we would not
automatically pick up the reverse device mapping.

As a result, users had to specify both forward and backward device mapping
which is very tedious to do.

In this PR, I've added this functionality such that TensorPipe automatically
picks up the reverse device mapping during the backward pass. This is done by
storing the appropriate device mapping in the "recv" autograd function for
distributed autograd.

#Closes: https://github.com/pytorch/pytorch/issues/44170

Differential Revision: [D23751975](https://our.internmc.facebook.com/intern/diff/D23751975/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23751975/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Support device map for distributed autograd while using TensorPipe."


TensorPipe's `set_device_map` option was applied during the forward
pass. However, if we ran the backward pass for the graph we would not
automatically pick up the reverse device mapping.

As a result, users had to specify both forward and backward device mapping
which is very tedious to do.

In this PR, I've added this functionality such that TensorPipe automatically
picks up the reverse device mapping during the backward pass. This is done by
storing the appropriate device mapping in the "recv" autograd function for
distributed autograd.

#Closes: https://github.com/pytorch/pytorch/issues/44170

Differential Revision: [D23751975](https://our.internmc.facebook.com/intern/diff/D23751975/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23751975/)!

[ghstack-poisoned]

lw reviewed (commented):
This PR will conflict with some work @mrshenli is doing in #44418 and the stack @beauby is about to merge.

Also, wondering why you went the route of adding a per-RPC override of the device map, rather than for example the solution I proposed in https://github.com/pytorch/pytorch/issues/44170#issuecomment-688345529. This PR certainly introduces a more flexible approach, but which also comes with extra complexity, so I'm thinking whether for the initial version we should go for something simpler?

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -157,7 +157,9 @@ class TORCH_API RpcAgent {
       Message&& message,
-      const float rpcTimeoutSeconds = kUnsetRpcTimeout) = 0;
+      const float rpcTimeoutSeconds = kUnsetRpcTimeout,
+      const std::unordered_map<c10::DeviceIndex, c10::DeviceIndex>& deviceMap =
```
The only user facing API we support today is the Python one. The RpcAgent interface can be thought of as an internal API that we have complete control over. In this PR we do attach this map to the message and actually reverse it for distributed autograd.

I went with DeviceIndex here to be consistent with the rest of the device mapping code. I agree with Shen that this should be Device, but that is a much more involved change for 1.7. We control this interface and all its implementations, so it shouldn't be a big deal to change this parameter slightly in the future.

pritamdamania87 replied:
I don't think this is tied with whether or not we want to support per-RPC device map arguments. This is not the public API that users see and is a private one for now. If we do end up building a C++ API, at that point we can evaluate what to do with this extra argument.

Regarding complexity I'm not sure if there is a simpler way to address this issue holistically. A global map for the backward wouldn't work in all cases. For example if nodes 1 and 2 perform RPCs on node3 with separate global device maps for the forward pass, there can't be a global backward map defined on node3 to handle this. It seems like we do need to do this at a per RPC level to handle this in a generic way.

pritamdamania87 commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -43,6 +43,23 @@ inline void checkCPUTensor(const torch::Tensor& tensor) {
+    const auto deviceIter = deviceMap.find(tensor.device().index());
+    if (deviceIter == deviceMap.end()) {
+      checkCPUTensor(tensor);
+      deviceIndices.push_back(-1);
```
Not sure I completely followed the idea here, but downstream code like tensorpipeSerialize rely on deviceIndices being the same length as tensors, so we do need to put in some placeholder deviceIndex for CPU tensors here?

pritamdamania87 commented:
> This PR will conflict with some work @mrshenli is doing in #44418 and the stack @beauby is about to merge.

We can probably resolve the conflicts based on the order in which the PRs land.

> Also, wondering why you went the route of adding a per-RPC override of the device map, rather than for example the solution I proposed in #44170 (comment). This PR certainly introduces a more flexible approach, but which also comes with extra complexity, so I'm thinking whether for the initial version we should go for something simpler?

I felt it wasn't too hard to implement the flexible approach and that's why I went with this approach. In the long term we probably want to have the backward pass pick up the device mapping like this anyways and since it wasn't too much work in implementing this, I thought it would be better to just go with this approach from the get go.

pritamdamania committed:
Update on "Support device map for distributed autograd while using TensorPipe."


TensorPipe's `set_device_map` option was applied during the forward
pass. However, if we ran the backward pass for the graph we would not
automatically pick up the reverse device mapping.

As a result, users had to specify both forward and backward device mapping
which is very tedious to do.

In this PR, I've added this functionality such that TensorPipe automatically
picks up the reverse device mapping during the backward pass. This is done by
storing the appropriate device mapping in the "recv" autograd function for
distributed autograd.

#Closes: https://github.com/pytorch/pytorch/issues/44170

Differential Revision: [D23751975](https://our.internmc.facebook.com/intern/diff/D23751975/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23751975/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Support device map for distributed autograd while using TensorPipe."


TensorPipe's `set_device_map` option was applied during the forward
pass. However, if we ran the backward pass for the graph we would not
automatically pick up the reverse device mapping.

As a result, users had to specify both forward and backward device mapping
which is very tedious to do.

In this PR, I've added this functionality such that TensorPipe automatically
picks up the reverse device mapping during the backward pass. This is done by
storing the appropriate device mapping in the "recv" autograd function for
distributed autograd.

#Closes: https://github.com/pytorch/pytorch/issues/44170

Differential Revision: [D23751975](https://our.internmc.facebook.com/intern/diff/D23751975/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23751975/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Support device map for distributed autograd while using TensorPipe."


TensorPipe's `set_device_map` option was applied during the forward
pass. However, if we ran the backward pass for the graph we would not
automatically pick up the reverse device mapping.

As a result, users had to specify both forward and backward device mapping
which is very tedious to do.

In this PR, I've added this functionality such that TensorPipe automatically
picks up the reverse device mapping during the backward pass. This is done by
storing the appropriate device mapping in the "recv" autograd function for
distributed autograd.

#Closes: https://github.com/pytorch/pytorch/issues/44170

Differential Revision: [D23751975](https://our.internmc.facebook.com/intern/diff/D23751975/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23751975/)!

[ghstack-poisoned]

pritamdamania87 commented:
@mrshenli @lw We didn't get this into 1.7 since device maps were reverted for that release, although I feel we should land this in 1.8. Let me know your thoughts on this and if it makes sense, would love to have another round of review :)

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -646,8 +684,15 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
+        getDevicesForRemote(clientPipe.pipe_->getRemoteName(), requestMessage);
+  } else {
+    // If deviceMap is specified, use that instead.
+    devices = getDevicesForTensors(
```
Is it possible to consolidate `getDevicesForTensors` and `getDevicesForRemote` into one by letting it take an optional device map arg?

pritamdamania87 replied:
`getDevicesForTensors` is actually a subset of `getDevicesForRemote`. `getDevicesForRemote` internally calls `getDevicesForTensors` to avoid any duplication in logic.

mrshenli replied:
Yep, I noticed that. What I didn't follow is why we need two similar but slightly different utility functions, and their didn't convey their difference? If all we need to letting the second utility function skip the map search and use the provided device map, this can potentially be done by using an optional arg? But this is a nit, please feel free to ignore.

mrshenli reviewed (commented):
LGTM! Shall we wait for #44418 before landing this one? Because if that one didn't land in time, this one will also needs to be reverted. And #44418 will need another fix in TensorPipe to detect GPUs that does not support peer access, i.e., `cudaErrorPeerAccessUnsupported`. I will wait that fix in TP, and then rebase and land  #44418. cc @lw @beauby

pritamdamania87 commented:
Sure, we can wait for #44418

pritamdamania committed:
Update on "Support device map for distributed autograd while using TensorPipe."


TensorPipe's `set_device_map` option was applied during the forward
pass. However, if we ran the backward pass for the graph we would not
automatically pick up the reverse device mapping.

As a result, users had to specify both forward and backward device mapping
which is very tedious to do.

In this PR, I've added this functionality such that TensorPipe automatically
picks up the reverse device mapping during the backward pass. This is done by
storing the appropriate device mapping in the "recv" autograd function for
distributed autograd.

#Closes: https://github.com/pytorch/pytorch/issues/44170

Differential Revision: [D23751975](https://our.internmc.facebook.com/intern/diff/D23751975/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23751975/)!

[ghstack-poisoned]

pritamdamania87 commented:
@mrshenli Could you take another look? Thanks!

mrshenli reviewed (approved):
LGTM! Thanks for adding this!
